dipy/denoise/randommatrix_localpca_parallel.pyx

```diff
@@ -0,0 +1,301 @@
+#!python
+
+cimport cython
+cimport safe_openmp as openmp
+
+import numpy as np
+import time
+from safe_openmp cimport have_openmp
+from cython.parallel import prange
+from multiprocessing import cpu_count
+from libc.math cimport sqrt
+from math import floor
+from scipy.linalg.cython_blas cimport dgemm
+from scipy.linalg.cython_blas cimport dgemv
+from scipy.linalg.cython_lapack cimport dsyevd
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef int fast_matvec(char ta, double[:,::1] A, double[:] b, double[:] y, double alpha=1.0, double beta=0.0,int incx=1) nogil except -1:
+
+    cdef:
+        char transa
+        int m,n
+        double *a0=&A[0,0]
+        double *b0=&b[0]
+        double *y0=&y[0]
+ 
+
+    if ta == 'n':
+        transa='n'        
+        n= A.shape[0]
+        m= A.shape[1]
+ 
+        dgemv(&transa, &m , &n, &alpha, a0, &m,  b0, &incx, &beta, y0, &incx)
+    else:
+        transa='t'        
+        n= A.shape[0]
+        m= A.shape[1]
+
+
+        dgemv(&transa, &m , &n, &alpha, a0, &m,  b0, &incx, &beta, y0, &incx)
+
+    return 0
+
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+cpdef int fast_eig(double[:,::1] a, double[::1] W, double[::1] WORK, int LWORK, int[::1] IWORK, int LIWORK) nogil except -1:
+
+    cdef:
+        char JOBZ='V' 
+        char UPLO='U'
+        int N = a.shape[0]
+        double *a0=&a[0,0]
+        double *w0=&W[0]
+        double *work0=&WORK[0]
+        int *iwork0=&IWORK[0]
+
+        int lda=N
+        int lw=LWORK
+        int liw= LIWORK
+        int info
+
+    dsyevd( &JOBZ, &UPLO, &N, a0, &lda, w0,work0,&LWORK, iwork0,&liw,&info)
+    return 0
+
+
+
+@cython.boundscheck(False)  # Deactivate bounds checking
+@cython.wraparound(False)   # Deactivate negative indexing.
+@cython.nonecheck(False)
+cpdef int fast_dgemm(double[:,::1] a, double[:,::1] c) nogil except -1:
+
+    cdef:
+        char transa 
+        char transb
+        int m2, n2, m,n,k
+        double alpha=1.0
+        double beta=0.0
+        double *a0=&a[0,0]
+        double *c0=&c[0,0]
+
+    m2= a.shape[0]         
+    n2= a.shape[1]
+
+    if m2 <= n2:           # XXt
+        transa='t'
+        transb='n'  
+        m=m2
+        n=m2
+        k=n2
+         
+        dgemm(&transa, &transb, &m , &n, &k, &alpha, a0, &k, a0,
+               &k, &beta, c0, &m)  
+    else:                # XtX
+        transa='n'
+        transb='t'
+        m=n2
+        n=n2
+        k=m2
+         
+        dgemm(&transa, &transb, &m , &n, &k, &alpha, a0, &m, a0,
+               &n, &beta, c0, &m)  
+       
+    return 0
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.nonecheck(False)
+def randommatrix_localpca_parallel(arr, patch_extent=0, out_dtype=None,num_threads=None):
```

Comment:
Can you remake this function to something shorter? Suggestion: pca_denoise.

Reply:
Hi Anh! I agree with @RafaelNH on this one, I would suggest randommatrix_lpca.pyx